code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -1664,85 +1663,101 @@ fn update_rewards_pool<T: Config>(
 ) -> Result<(), RewardAccumulationCalculationError> {
-	if reward.reward_rate.amount.is_zero() {
+	// short-circuit if the reward rate amount is zero
+	let Some(reward_rate_amount) = NonZeroU128::new(reward.reward_rate.amount.into()) else {
 		return Ok(())
-	}
+	};
 
 	let elapsed_time = now_seconds
 		.safe_sub(&reward.last_updated_timestamp)
 		.map_err(|_| RewardAccumulationCalculationError::BackToTheFuture)?;
 
 	let reward_rate_period_seconds = reward.reward_rate.period.as_secs();
 
-	// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
-	// reward_rate_period_seconds is a NonZeroU64 here
-	let periods_surpassed = elapsed_time.div(reward_rate_period_seconds.get());
+	let periods_surpassed = <u64 as Div<NonZeroU64>>::div(elapsed_time, reward_rate_period_seconds);
 
-	if periods_surpassed.is_zero() {
-		Ok(())
-	} else {
-		let total_locked_rewards: u128 = T::Assets::balance_on_hold(asset_id, pool_account).into();
-
-		// the maximum amount repayable given the reward rate.
-		// i.e. if total locked is 50, and the reward rate is 15, then this would be 3
-		//
-		// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
-		// reward.reward_rate.amount is known to be non-zero here as per the check at the beginning
-		// of this function
-		let maximum_releasable_periods = total_locked_rewards.div(reward.reward_rate.amount.into());
-
-		let releasable_periods_surpassed =
-			cmp::min(maximum_releasable_periods, periods_surpassed.into());
-
-		// SAFETY: Usage of mul is safe here because newly_accumulated_rewards =
-		//      (    total_locked_rewards            elapsed_time        )
-		//  min ( ------------------------- , -------------------------- )
-		//      ( reward.reward_rate.amount   reward_rate_period_seconds )
-		// If LHS is smaller, this will result in total_locked_rewards, which we know fits in a
-		// u128. If RHS is smaller, this has to be < total_locked_rewards, so it will also fit.
-		let newly_accumulated_rewards =
-			releasable_periods_surpassed.mul(reward.reward_rate.amount.into());
-		let new_total_rewards = newly_accumulated_rewards
-			.safe_add(&reward.total_rewards.into())
-			.map_err(|_| RewardAccumulationCalculationError::ArithmeticError)?;
-
-		// u64::MAX is roughly 584.9 billion years in the future, so saturating at that should be ok
-		let last_updated_timestamp = reward.last_updated_timestamp.saturating_add(
-			reward_rate_period_seconds
-				.get()
-				.saturating_mul(releasable_periods_surpassed.saturated_into::<u64>()),
-		);
-
-		if !newly_accumulated_rewards.is_zero() {
-			T::Assets::release(
-				asset_id,
-				pool_account,
-				newly_accumulated_rewards.into(),
-				false, // not best effort, entire amount must be released
-			)
-			.expect("funds should be available to release based on previous check; qed;");
+	// if no periods have been surpassed, short-circuit
+	let Some(periods_surpassed) = NonZeroU64::new(periods_surpassed) else {
```

Comment:
Excessive verbosity to check if something is zero and return no?

Reply:
This also carries the non-zero invariant through the rest of the code without having to track it ourselves